public/app/plugins/datasource/loki/datasource.ts

```diff
@@ -186,31 +188,20 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
     options: RangeQueryOptions,
     responseListLength = 1
   ): Observable<DataQueryResponse> => {
-    // target.maxLines value already preprocessed
-    // available cases:
-    // 1) empty input -> mapped to NaN, falls back to dataSource.maxLines limit
-    // 2) input with at least 1 character and that is either incorrect (value in the input field is not a number) or negative
-    //    - mapped to 0, falls back to the limit of 0 lines
-    // 3) default case - correct input, mapped to the value from the input field
-
-    let linesLimit = 0;
-    if (target.maxLines === undefined) {
-      // no target.maxLines, using options.maxDataPoints
-      linesLimit = Math.min((options as DataQueryRequest<LokiQuery>).maxDataPoints || Infinity, this.maxLines);
-    } else {
-      // using target.maxLines
-      if (isNaN(target.maxLines)) {
-        linesLimit = this.maxLines;
-      } else {
-        linesLimit = target.maxLines;
-      }
-    }
+    // For metric query we use maxDataPoints from the request options which should be something like width of the
+    // visualisation in pixels. In case of logs request we either use lines limit defined in the query target or
+    // global limit defined for the data source which ever is lower.
+    let maxDataPoints = isMetricsQuery(target.expr)
+      ? // We fallback to maxLines here because maxDataPoints is defined as possibly undefined. Not sure that can
+        // actually happen both Dashboards and Explore should send some value here. If not maxLines does not make that
+        // much sense but nor any other arbitrary value.
+        (options as DataQueryRequest<LokiQuery>).maxDataPoints || this.maxLines
+      : Math.min(target.maxLines || Infinity, this.maxLines);
```

Comment:
So this mimicks the current behaviour but I wonder if it makes sense. Shouldn't user be able to override the global limit from the query editor in both directions? Ie rise the limit for a particular query if needed? @davkal

Reply:
No, I don't think it makes sense. Because with logs query, we want to either use the `this.maxLines` if users didn't fill the query row Line limit, or if they filled it, then use that value.  So I think it should be something like `target.maxLines || this.maxLines;`. Not sure what should be the behaviour when user selects 0 in query row's Line limit - if showing 0 logs or showing default number of logs.

Reply:
I agree it looks weird at the moment. Will change it.